### PR TITLE
Emit diagnostics for device lambdas

### PR DESCRIPTION
**Pull request by senior-zero**

### CODE REVIEW

**Comment by senior-zero** [comment]
run tests

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
We shouldn't use CUB headers outside of `thrust/system/cuda/`, as CUB is only a dependency when the CUDA backend is enabled.

Let's replicate the logic from CUB in Thrust's type_traits.h and use that instead.

**Comment by jrhemstad** [comment]
All the logic is in `cuda::std::invoke_result` isn't it? Can't we just use `cuda::std::invoke_result` directly?

**Comment by senior-zero** [comment]
> All the logic is in `cuda::std::invoke_result` isn't it? Can't we just use `cuda::std::invoke_result` directly?

Invoke result is C++17 feature. CUB wrapper switches to result of for previous dialects.

**Comment by jrhemstad** [comment]
> Invoke result is C++17 feature. CUB wrapper switches to result of for previous dialects.

Can't we back port `cuda::std::invoke_result_t` to be available prior to C++17? 

It's just a library feature, and we already have back ported many Standard Library features in libcu++, e.g., https://github.com/NVIDIA/libcudacxx/pull/66

**Comment by allisonvacanti** [comment]
> Can't we back port cuda::std::invoke_result_t to be available prior to C++17?

We'd need a libcu++ release by 8/15 to do this.

File: thrust/iterator/detail/transform_input_output_iterator.inl [hunk 1]
```diff
@@ -35,12 +36,7 @@ template <typename InputFunction, typename OutputFunction, typename Iterator>
 {
   using iterator_value_type = typename thrust::iterator_value<Iterator>::type;
 
-  // std::result_of is deprecated in 2017, replace with std::invoke_result
-#if THRUST_CPP_DIALECT < 2017
-  using Value = typename std::result_of<InputFunction(iterator_value_type)>::type;
-#else
-  using Value = std::invoke_result_t<InputFunction, iterator_value_type>;
-#endif
+  using Value = cub::detail::invoke_result_t<InputFunction, iterator_value_type>;
```
Comment (miscco) [inline thrust/iterator/detail/transform_input_output_iterator.inl:39]:
```suggestion
  using Value = CUB_NS_QUALIFIER::detail::invoke_result_t<InputFunction, iterator_value_type>;
```

File: thrust/iterator/detail/transform_input_output_iterator.inl [hunk 2]
```diff
@@ -93,12 +89,7 @@ public:
     <
         transform_input_output_iterator<InputFunction, OutputFunction, Iterator>
       , Iterator
-    // std::result_of is deprecated in 2017, replace with std::invoke_result
-#if THRUST_CPP_DIALECT < 2017
-      , typename std::result_of<InputFunction(iterator_value_type)>::type
-#else
-      , std::invoke_result_t<InputFunction, iterator_value_type>
-#endif
+      , cub::detail::invoke_result_t<InputFunction, iterator_value_type>
```
Comment (miscco) [inline thrust/iterator/detail/transform_input_output_iterator.inl:92]:
```suggestion
      , CUB_NS_QUALIFIER::detail::invoke_result_t<InputFunction, iterator_value_type>
```

**Comment by senior-zero** [comment]
run tests

**Review by allisonvacanti — APPROVED** [review]

**Pull request merged**